man/profile.conf.5
.TH docker4ssh.conf 5 "December 13, 2021" docker4ssh.conf "docker4ssh configuration file"

.SH SYNOPSIS
.TP
/etc/docker4ssh/profile/*

.SH SECTION NAME
.TP
A representative name for the profile

.SH KEYS
\fBUsername\fR = username
Username for this profile.
The username can be specified as plain text or regex:
    Regex: Put \fIregex:\fR in front of it. The regex must must be \fBgo\fR / \fBgolang\fR compatible. Visit \fIregex101.com\fR to validate your regex.
.TP

.TP
\fBPassword\fR = password
Password for the profile.
The password can be specified as plain text, regex or hash:
    Regex: Put \fIregex:\fR in front of it. The regex must must be \fBgo\fR / \fBgolang\fR compatible. Visit \fIregex101.com\fR to validate your regex.
    Hash: Put \fIsha1:\fR, \fIsha256:\fR or \fIsha512:\fR in front of it. Note that the hash must be hashed with the prefix algorithm.
.TP

\fBNetworkMode\fR = 1 | 2 | 3 | 4 | 5
Default network mode for every connection.
NetworkMode describes the behavior of the container's network
Must be one of the following:
    1 (Off): Disable networking complete.
    2 (Isolate): Isolates the container from the host and the host's network. Therefore, no configurations can be changed from within the container.
    3 (Host): Default docker network.
    4 (Docker): Same as \fI3\fR but the container is in a docker4ssh controlled subnet. This is useful to differ normal from docker4ssh containers.
    5 (None): disables all isolation between the docker container and the host, so inside the network the container can act as the host. So it has access to the host's network directly.
.TP

\fBConfigurable\fR = true | false
Default configurable setting for every connection.
Configurable describes if the container should be configurable from within it. This means that the connect user is able to change all settings which are described here.
Must be true or false.
.TP

\fBRunLevel\fR = 1 | 2 | 3
Default run level for every connection.
RunRevel describes the container behavior when the user connection to a container is stopped.
Must be one of the following:
    1 (User): The container stops working if no user is connected to it anymore.
    2 (Container): The container runs when no user is connected \fIExitAfter\fR is specified.
    3 (Forever): The container runs forever.
.br
Note that the container exits always, independent of its RunLevel, when the via \fIExitAfter\fR specified process ends.
.TP

\fBStartupInformation\fR = true | false
Default startup information setting for every connection.
StartupInformation specifies if information about the container (id, network mode, ...) should be shown when a user connects to it.
Must be true or false.
.TP

\fBExitAfter\fR = exit after
Default exit after process for every process.
ExitAfter is a process name after which end the container should stop running.
.TP

\fBKeepOnExit\fR = true | false
Default keep on exit setting for every connection.
KeepOnExit specifies if the container should be saved when it stops working.
Must be true or false.

.SH EXAMPLE
[test]
.br
Username = "test"
.br
Image = "alpine:latest"

.SH SEE ALSO
docker4ssh(1), docker4ssh.conf(5)

.SH AUTHORS
Written by ByteDream (https://github.com/ByteDream)


server/config/profile.go
package config

import (
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/json"
	"fmt"
	"github.com/BurntSushi/toml"
	"go.uber.org/zap"
	"hash"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type Profile struct {
	name               string
	Username           *regexp.Regexp
	Password           *regexp.Regexp
	passwordHashAlgo   hash.Hash
	NetworkMode        int
	Configurable       bool
	RunLevel           int
	StartupInformation bool
	ExitAfter          string
	KeepOnExit         bool
	Image              string
	ContainerID        string
}

func (p *Profile) Name() string {
	return p.name
}

func (p *Profile) Match(user string, password []byte) bool {
	// username should only be nil if profile was generated from Config.Profile.Dynamic
	if p.Username == nil || p.Username.MatchString(user) {
		if p.passwordHashAlgo != nil {
			password = p.passwordHashAlgo.Sum(password)
		}
		return p.Password.Match(password)
	}
	return false
}

type preProfile struct {
	Username           string
	Password           string
	NetworkMode        int
	Configurable       bool
	RunLevel           int
	StartupInformation bool
	ExitAfter          string
	KeepOnExit         bool
	Image              string
	Container          string
}

func LoadProfileFile(path string, defaultPreProfile preProfile) (Profiles, error) {
	var rawProfile map[string]interface{}
	if _, err := toml.DecodeFile(path, &rawProfile); err != nil {
		return nil, err
	}

	profiles, err := parseRawProfile(rawProfile, path, defaultPreProfile)
	if err != nil {
		return nil, err
	}

	return profiles, nil
}

func LoadProfileDir(path string, defaultPreProfile preProfile) (Profiles, error) {
	dir, err := os.ReadDir(path)
	if err != nil {
		return nil, err
	}

	var profiles Profiles
	for i, profileConf := range dir {
		p, err := LoadProfileFile(filepath.Join(path, profileConf.Name()), defaultPreProfile)
		if err != nil {
			return nil, err
		}
		profiles = append(profiles, p...)
		zap.S().Debugf("Pre-loaded file %d (%s) with %d profile(s)", i+1, profileConf.Name(), len(p))
	}

	return profiles, nil
}

func parseRawProfile(rawProfile map[string]interface{}, path string, defaultPreProfile preProfile) (profiles []*Profile, err error) {
	var count int
	for key, value := range rawProfile {
		rawValue, err := json.Marshal(value)
		if err != nil {
			return nil, err
		}
		pp := preProfile{
			NetworkMode:        3,
			RunLevel:           1,
			StartupInformation: true,
		}
		if err = json.Unmarshal(rawValue, &pp); err != nil {
			return nil, fmt.Errorf("failed to parse %s profile conf file %s: %v", key, path, err)
		}

		var rawUsername string
		if rawUsername = strings.TrimPrefix(pp.Username, "regex:"); rawUsername == pp.Username {
			rawUsername = strings.ReplaceAll(rawUsername, "*", ".*")
		}
		if !strings.HasSuffix(rawUsername, "$") {
			rawUsername += "$"
		}
		username, err := regexp.Compile("(?m)" + rawUsername)
		if err != nil {
			return nil, fmt.Errorf("failed to parse %s profile username regex for conf file %s: %v", key, path, err)
		}

		var rawPassword string
		if rawPassword = strings.TrimPrefix(pp.Password, "regex:"); rawUsername == pp.Password {
			rawPassword = strings.ReplaceAll(rawPassword, "*", ".*")
		}
		algo, rawPasswordOrHash := getHash(rawPassword)
		if algo == nil && rawPasswordOrHash == "" {
			rawPasswordOrHash = ".*"
		}
		if !strings.HasSuffix(rawPasswordOrHash, "$") {
			rawPasswordOrHash += "$"
		}
		password, err := regexp.Compile("(?m)" + rawPasswordOrHash)
		if err != nil {
			return nil, fmt.Errorf("failed to parse %s profile password regex for conf file %s: %v", key, path, err)
		}

		if (pp.Image == "") == (pp.Container == "") {
			return nil, fmt.Errorf("failed to interpret %s profile image / container definition for conf file %s: `Image` or `Container` must be specified, not both nor none of them", key, path)
		}

		profiles = append(profiles, &Profile{
			name:               key,
			Username:           username,
			Password:           password,
			passwordHashAlgo:   algo,
			NetworkMode:        pp.NetworkMode,
			Configurable:       pp.Configurable,
			RunLevel:           pp.RunLevel,
			StartupInformation: pp.StartupInformation,
			ExitAfter:          pp.ExitAfter,
			KeepOnExit:         pp.KeepOnExit,
			Image:              pp.Image,
			ContainerID:        pp.Container,
		})
		count++
		zap.S().Debugf("Pre-loaded profile %s (%d)", key, count)
	}
	return
}

type Profiles []*Profile

func (ps Profiles) GetByName(name string) (*Profile, bool) {
	for _, profile := range ps {
		if profile.name == name {
			return profile, true
		}
	}
	return nil, false
}

func (ps Profiles) Match(user string, password []byte) (*Profile, bool) {
	for _, profile := range ps {
		if profile.Match(user, password) {
			return profile, true
		}
	}
	return nil, false
}

func DefaultPreProfileFromConfig(config *Config) preProfile {
	defaultProfile := config.Profile.Default

	return preProfile{
		Password:           defaultProfile.Password,
		NetworkMode:        defaultProfile.NetworkMode,
		Configurable:       defaultProfile.Configurable,
		RunLevel:           defaultProfile.RunLevel,
		StartupInformation: defaultProfile.StartupInformation,
		ExitAfter:          defaultProfile.ExitAfter,
		KeepOnExit:         defaultProfile.KeepOnExit,
	}
}

func HardcodedPreProfile() preProfile {
	return preProfile{
		NetworkMode:        3,
		RunLevel:           1,
		StartupInformation: true,
	}
}

func DynamicProfileFromConfig(config *Config, defaultPreProfile preProfile) (Profile, error) {
	raw, err := json.Marshal(config.Profile.Dynamic)
	if err != nil {
		return Profile{}, err
	}
	json.Unmarshal(raw, &defaultPreProfile)

	algo, rawPasswordOrHash := getHash(defaultPreProfile.Password)
	if algo == nil && rawPasswordOrHash == "" {
		rawPasswordOrHash = ".*"
	}
	password, err := regexp.Compile("(?m)" + rawPasswordOrHash)
	if err != nil {
		return Profile{}, fmt.Errorf("failed to parse password regex: %v ", err)
	}

	return Profile{
		name:               "",
		Username:           nil,
		Password:           password,
		passwordHashAlgo:   algo,
		NetworkMode:        defaultPreProfile.NetworkMode,
		Configurable:       defaultPreProfile.Configurable,
		RunLevel:           defaultPreProfile.RunLevel,
		StartupInformation: defaultPreProfile.StartupInformation,
		ExitAfter:          defaultPreProfile.ExitAfter,
		KeepOnExit:         defaultPreProfile.KeepOnExit,
	}, nil
}

func getHash(password string) (algo hash.Hash, raw string) {
	split := strings.SplitN(password, ":", 1)

	if len(split) == 1 {
		return nil, password
	} else {
		raw = split[1]
	}

	switch split[0] {
	case "sha1":
		algo = sha1.New()
	case "sha256":
		algo = sha256.New()
	case "sha512":
		algo = sha512.New()
	default:
		algo = nil
	}
	return
}
